docs/tutorial.rst

```diff
@@ -0,0 +1,1842 @@
+This is a standard benchmark for measuring the DRAM bandwidth.
+The tutorial container already contains a pre-compiled OpenMP version of the benchmark.
+Our test will run the STREAM executable, validate the output and extract the figure of merits.
+Here is the full ReFrame test:
+
+.. literalinclude:: ../examples/tutorial/stream/stream_runonly.py
+   :caption:
+   :pyobject: stream_test
+
+ReFrame tests are specially decorated classes that ultimately derive from the :class:`~reframe.core.pipeline.RegressionTest` class.
+Since we only want to run an executable in this first test, we derive from the :class:`~reframe.core.pipeline.RunOnlyRegressionTest` class, which essentially short-circuits the "compile" stage of the test.
+The :func:`@simple_test <reframe.core.decorators.simple_test>` decorator registers a test class with the framework and makes it available for running.
+
+Every ReFrame test must define the :attr:`valid_systems` and :attr:`valid_prog_environs` variables.
+These describe the test's constraints and the framework will automatically filter out tests on systems and environments that do not match the constraints.
+We will describe the system and environments abstractions later in this tutorial.
+For this first example, the ``*`` symbol denotes that this test is valid for any system or environment.
+A :class:`~reframe.core.pipeline.RunOnlyRegressionTest` must also define an :attr:`executable` to run.
+
+A test must also define a validation function which is decorated with the :func:`@sanity_function<reframe.core.pipeline.RegressionMixin.sanity_function>` decorator.
+This function will be used to validate the test's output after it is finished.
+ReFrame, by default, makes no assumption about whether a test is successful or not;
+it is the test's responsibility to define its validation.
+The framework provides a rich set of `utility functions <deferrable_functions_reference>`__ that help matching patterns and extract values from the test's output.
+The :attr:`stdout` here refers to the name of the file where the test's standard output is stored.
+
+Finally, a test may optionally define a set of *performance functions* that will extract *figures of merit* for the test.
+These are simple test methods decorated with the :func:`@performance_function <reframe.core.builtins.performance_function>` decorator that return the figure of merit.
+In this example, we extract the ``Copy`` and ``Triad`` bandwidth values and convert them to ``float``.
+These figures of merit or *performance variables* as they are called in ReFrame's nomenclature have a special treatment:
+they are logged in the test's performance logs and a reference value per system may also be assigned to them.
+If that reference value is not met within some user-defined thresholds, the test will fail.
+
+
+Running a test
+==============
+
+Running our test is very straightforward:
+
+.. code-block:: bash
+
+   cd reframe-examples/tutorial
+   reframe -c stream/stream_runonly.py -r
```

Comment:
Even as a `RunOnlyRegressionTest`, this assumes a binary `stream.x` exists, which is not the case out of the box.

Reply:
Yes, but it is inside the container. I didn't want to introduce right away the full build and run process, as this would require to introduce the notion of systems and environments. The first sentence in "Systems and environments" explains that up to that point we have been working with the assumption that "stream.x" was available and then I'm introducing the concepts + test constraints

Reply:
In my case it cannot find stream.x either. It is built using a Dockerfile, but probably the above test should target that node.

Reply:
I did a fresh build of the container and it works for me. Could you retry?

Reply:
I think the best is to add caption to every command listing denoting the container to use.